### PR TITLE
Node and project nodes move & rename, project deletion

### PR DESCRIPTION
**Pull request by nashousni**
Drag and drop added to the GUI

### CODE REVIEW

File: gse-app/src/main/java/com/powsybl/gse/app/ProjectPane.java [hunk 1]
```diff
@@ -38,11 +38,17 @@
...
+import javafx.scene.input.TransferMode;
+import javafx.scene.input.ClipboardContent;
+
 /**
  * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
  */
 public class ProjectPane extends Tab {
-
+    Object source; //ajouté
+    TreeItem sourceTreeItem;
```
Comment (geofjamg) [inline gse-app/src/main/java/com/powsybl/gse/app/ProjectPane.java]: Encapsulate these 3 fields in a MoveContext object.
Add private  modifier

File: gse-app/src/main/java/com/powsybl/gse/app/ProjectPane.java [hunk 2]
```diff
@@ -203,6 +209,63 @@ protected void updateItem(Object value, boolean empty) {
...
+                                        cb.putString(((ProjectFolder) value).getName());
+                                    }
+                                    db.setContent(cb);
+                                    event.consume();
+                                }
+                            });
+                            if (node instanceof ProjectFolder) {
+                                ProjectFolder projectFolder = (ProjectFolder) node;
+                                setOnDragOver(event -> {
+                                    if (event.getGestureSource() != this && event.getDragboard().hasString()) {  //à modifier hasString
```
Comment (geofjamg) [inline gse-app/src/main/java/com/powsybl/gse/app/ProjectPane.java]: => english

File: gse-app/src/main/java/com/powsybl/gse/app/ProjectPane.java [hunk 3]
```diff
@@ -203,6 +209,63 @@ protected void updateItem(Object value, boolean empty) {
...
+                                        event.acceptTransferModes(TransferMode.ANY);
+
+                                    }
+                                    event.consume();
+
+                                });
+                                setOnDragDropped(event -> {
+                                    //getItem récupère l'objet , getTreeItem récupère le treeItem
```
Comment (geofjamg) [inline gse-app/src/main/java/com/powsybl/gse/app/ProjectPane.java]: => english

File: gse-app/src/main/java/com/powsybl/gse/app/ProjectPane.java [hunk 4]
```diff
@@ -203,6 +209,63 @@ protected void updateItem(Object value, boolean empty) {
...
+                                    TreeItem treeItemaccepteur = getTreeItem();
+                                    System.out.println("Bienvenue dans le dossier " + getItem().toString().toUpperCase());
```
[the remark this answers is not in the document]
Comment (geofjamg) [reply gse-app/src/main/java/com/powsybl/gse/app/ProjectPane.java]: To remove, this is a debug log or use slf4j logger instead
=> english

File: gse-app/src/main/java/com/powsybl/gse/app/ProjectPane.java [hunk 5]
```diff
@@ -203,6 +209,63 @@ protected void updateItem(Object value, boolean empty) {
...
+                                    Dragboard db = event.getDragboard();
+
+                                    boolean success = false;
+                                    if (db.hasString()) {
+                                        if (source instanceof ProjectNode) {
+                                            ProjectNode monfichier = (ProjectNode) source;
+                                            monfichier.moveTo(projectFolder);
+
+                                            refresh(sourceparentTreeItem);
+                                            refresh(treeItemaccepteur);  //B
```
Comment (geofjamg) [inline gse-app/src/main/java/com/powsybl/gse/app/ProjectPane.java]: //B what does it mean?

File: gse-app/src/main/java/com/powsybl/gse/app/ProjectPane.java [hunk 6]
```diff
@@ -203,6 +209,63 @@ protected void updateItem(Object value, boolean empty) {
...
+
+                                            success = true;
+                                        }
+
+                                    } else {
+                                        System.out.println("rien");
```
Comment (geofjamg) [inline gse-app/src/main/java/com/powsybl/gse/app/ProjectPane.java]: To remove

File: gse-app/src/main/java/com/powsybl/gse/app/ProjectPane.java [hunk 7]
```diff
@@ -203,6 +216,119 @@ protected void updateItem(Object value, boolean empty) {
...
+                            setTextFill(Color.BLACK);
+                            setOpacity(1);
+                        } else if (value instanceof ProjectNode) {
+                            ProjectNode node = (ProjectNode) value;
+                            setText(node.getName());
+                            setGraphic(getTreeItem().getGraphic());
+                            setTextFill(Color.BLACK);
+                            setOpacity(node instanceof UnknownProjectFile ? 0.5 : 1);
+
+                            setOnDragDetected(event -> {
```
Comment (geofjamg) [inline gse-app/src/main/java/com/powsybl/gse/app/ProjectPane.java]: This callback is reset at each refresh of the treeview. This is a design (and performance) issue.

File: gse-app/src/main/java/com/powsybl/gse/app/ProjectPane.java [hunk 8]
```diff
@@ -203,6 +216,119 @@ protected void updateItem(Object value, boolean empty) {
...
+                                    } else if (value instanceof ProjectFolder) {
+                                        cb.putString(((ProjectFolder) value).getName());
+                                    }
+                                    db.setContent(cb);
+                                    event.consume();
+                                }
+                            });
+                            if (node instanceof ProjectFolder) {
+                                ProjectFolder projectFolder = (ProjectFolder) node;
+                                setOnDragOver(event -> {
```
Comment (geofjamg) [inline gse-app/src/main/java/com/powsybl/gse/app/ProjectPane.java]: Same remark

File: gse-app/src/main/java/com/powsybl/gse/app/ProjectPane.java [hunk 9]
```diff
@@ -203,6 +216,119 @@ protected void updateItem(Object value, boolean empty) {
...
+                                        if (count < 1) {
+                                            setTextFill(Color.CHOCOLATE);
+                                            setText(getText().toUpperCase());
+                                        }
+
+                                        event.acceptTransferModes(TransferMode.ANY);
+                                    }
+                                    event.consume();
+                                });
+                                setOnDragDropped(event -> {
```
Comment (geofjamg) [inline gse-app/src/main/java/com/powsybl/gse/app/ProjectPane.java]: Same remark

File: gse-app/src/main/java/com/powsybl/gse/app/ProjectPane.java [hunk 10]
```diff
@@ -203,6 +216,119 @@ protected void updateItem(Object value, boolean empty) {
...
+                                            ProjectNode monfichier = (ProjectNode) moveContext.source;
+                                            monfichier.moveTo(projectFolder);
+
+                                            refresh(moveContext.sourceparentTreeItem);
+                                            refresh(treeItemaccepteur);
+
+                                            success = true;
+                                        }
+
+                                    } else {
```
Comment (geofjamg) [inline gse-app/src/main/java/com/powsybl/gse/app/ProjectPane.java]: Empty else

File: gse-app/src/main/java/com/powsybl/gse/app/ProjectPane.java [hunk 11]
```diff
@@ -203,6 +216,119 @@ protected void updateItem(Object value, boolean empty) {
...
+                                    }
+                                    event.setDropCompleted(success);
+                                    event.consume();
+                                });
+                                setOnDragExited(event -> {
```
Comment (geofjamg) [inline gse-app/src/main/java/com/powsybl/gse/app/ProjectPane.java]: Same remark

File: gse-app/src/main/java/com/powsybl/gse/app/ProjectPane.java [hunk 12]
```diff
@@ -203,6 +216,119 @@ protected void updateItem(Object value, boolean empty) {
                     setGraphic(null);
                 } else {
                     fillCellInfosForObject(value);
+                    if (value == null) {
```
Comment (geofjamg) [inline gse-app/src/main/java/com/powsybl/gse/app/ProjectPane.java]: General comment pour updateItem. This method is too big. You should decompose into several methods.

File: gse-app/src/main/java/com/powsybl/gse/app/ProjectPane.java (hunk 7 above)
Comment (nashousni) [reply gse-app/src/main/java/com/powsybl/gse/app/ProjectPane.java]: //TreeItem Documentation
TreeItem is not a Node, and therefore no visual events will be fired on the TreeItem. To get these events, it is necessary to add relevant observers to the TreeCell instances (via a custom cell factory )

**Pull request merged**